### ISSUE TITLE
Bug: false negative in `id-match` rule

### ISSUE THREAD

**Issue by snitin315** [eslint/eslint#15443]
### Environment

Node version: v14.18.1
npm version: v8.1.4
Local ESLint version: v8.4.1 (Currently used)
Global ESLint version: v8.4.1
Operating System: darwin 20.6.0

### What parser are you using?

Default (Espree)

### [Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGlkLW1hdGNoOiBbXCJlcnJvclwiLCBcIl5bXmFdXCJdICovXG5cbmNvbnN0IGZvbyA9IHtcbiBbYV06IDEgLy8gc2hvdWxkIGJlIHJlcG9ydGVkXG59O1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)

<details>
<summary>Configuration</summary>

```
/* eslint id-match: ["error", "^[^a]"] */

```
</details>

```js
/* eslint id-match: ["error", "^[^a]"] */

const foo = {
 [a]: 1 // should be reported
};
```

### What did you expect to happen?

```js
/* eslint id-match: ["error", "^[^a]"] */

const foo = {
 [a]: 1 // Error: Identifier 'a' does not match the pattern '^[^a]'
};
```

### What actually happened?

```js
/* eslint id-match: ["error", "^[^a]"] */

const foo = {
 [a]: 1 // Lint free
```

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Refers https://github.com/eslint/eslint/pull/15431#discussion_r773074796.

**Comment by nzakas** [comment]
@snitin315 please be sure to move issues like this into “Ready to Implement” if you’re not looking for feedback. 

If you are looking for feedback, leave off the “accepted” label and place in the “Feedback Needed” column.
